passwordless-GDM: smartcard and passkey

ikerexxe opened the pull request:
Extension for the passwordless-GDM feature where the smartcard and passkey authentication methods have been added.

The design page is available at https://github.com/SSSD/sssd.io/pull/79.

You can use https://copr.fedorainfracloud.org/coprs/ipedrosa/passwordles-gdm/ for testing. As a reminder you should update sssd, mutter, gdm and gnome-shell packages and you also need to include the following configuration option in `/etc/sssd/sssd.conf`:
```
[pam]
pam_json_services = gdm-switchable-auth
```

Known limitations:
* If EIdP and passkey are enabled at the same time for a given user, from the two of them only EIdP will be advertised (other methods aren't affected). Changing it will affect performance, so it's been decided to leave it as it is.
* Since https://github.com/SSSD/sssd/pull/7983 isn't merged, SSSD is unable to propagate the `PIN request` and `PIN attempts left` information to GDM. This is temporary and once this PR or the other one are merged, I'll update the PR to make these functionality available.

ikerexxe committed:
Responder: generate JSON message for smartcard

Implement a set of functions to retrieve the smartcard data and generate
the JSON message with it.

Implement new unit test and adapt the existing ones to take into account
the new data.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: parse reply for smartcard

Parse GUI reply for smartcard and set the appropriate data in
`sss_auth_token` structure.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Add new options to tune smartcard prompting

These new options are needed by the GDM integration, but they can be
reused for CLI prompting.

:config: New options to tune smartcard prompting: 'init_prompt' and
         'pin_prompt'.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
sss_client: modify smartcard in prompt_config structure

Integration with GDM requests two prompts for smartcard so modifying the
prompt_config structure. In addition, implement all the functions needed
to manipulate the structure for these new prompts. Finally, add
unit-tests for the new functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
util: implement pam_get_response_data_all_same_type()

This API gets all the elements with the selected response type data from
the response_data linked list. Includes unit tests.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: check return value for json_string()

It returns NULL on error, but this wasn't checked.

Fixes: ceeffa9e1 ("Responder: generate JSON message for GUI")

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: refactor JSON functions to reduce args

Several of the functions in `pamsrv_json` had lots of arguments and I'm
about to add more for the passkey authentication mechanism. Reduce these
arguments by creating a structure that will contain all these data.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: generate JSON message for passkey

Implement a set of functions to retrieve the passkey data and generate
the JSON message with it.

Implement new unit test and adapt the existing ones to take into account
the new data.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Add new option to tune passkey PIN prompting

This new option is needed by the GDM integration, but it can be reused
for CLI prompting.

:config: New option to tune passkey prompting: 'pin_prompt'.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: parse reply for passkey

Parse GUI reply for passkey and set the appropriate data in
`sss_auth_token` structure.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
util: implement function to set passkey PIN

`sss_authtok_set_local_passkey_pin` provides a way to set the passkey
PIN in the authtok structure for local passkey authentication.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: set hardcoded strings for prompts

Due to the difficulty of having a single source for the prompts strings
for both CLI and GUI, it has been decided to leave them fixed and use
the strings proposed by Allan in the mockups design.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: extend smartcard JSON request message

Include the certificate data in the JSON messages to set it in the
authtok structure more easily.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: extend smartcard JSON reply message

Include the certificate data in the JSON message to set it in the
authtok structure more easily.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: make `decode_pam_passkey_msg()` public

This is needed by `pamsrv_json.c`, so let's make it public.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
krb5_child: advertise authentication methods

During the `preauthentication` phase krb5_child checks for the available
authentication methods for the given user, advertises them and the
process is kept alive. Once the state is change to `authentication` the
same krb5_child process processes the credentials and proceeds with the
authentication itself.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: update JSON message format

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -238,12 +237,11 @@ json_format_mechanisms(bool password_auth, const char *password_prompt,
-                                "init_prompt", oauth2_init_prompt,
-                                "link_prompt", oauth2_link_prompt,
+                                "initPrompt", oauth2_init_prompt,
+                                "linkPrompt", oauth2_link_prompt,
```
Any reason to prefix these variables with `oauth2_` but not `uri` or `code` ?

ikerexxe replied:
No, not really. In any case, this changed a bit in later commits, where I introduce the `auth_data` structure

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -549,9 +547,9 @@ json_unpack_auth_reply(struct pam_data *pd)
+    ret = json_unpack(jroot, "{s:o}", "authSelection", &jauth_selection);
     if (ret != 0) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "json_unpack for auth-selection failed.\n");
+        DEBUG(SSSDBG_CRIT_FAILURE, "json_unpack for authSelection failed.\n");
```
Maybe a macro for string "authSelection" makes sense here as I see you changed it here in 4 different places? Not necessary however.

ikerexxe replied:
It's not a bad idea, but I don't see it as necessary either. Whatever you prefer

justin-stephenson replied:
Feel free to ignore.

justin-stephenson commented on src/sss_client/pam_sss_prompt_config.c:
```diff
@@ -45,8 +45,9 @@ struct prompt_config_passkey {
-struct prompt_config_sc_pin {
-    char *prompt; /* Currently not used */
+struct prompt_config_smartcard {
+    char *prompt_init;
```
`prompt_inst` or `prompt_instruction`  to match more closely to what is mention in the design page?

ikerexxe replied:
I'm not sure I'm following you. Do you refer to the following text?
```
For clarification, there are two different types of user interaction messages:
`prompt` and `instruction`. The first one asks the user to enter something on
the screen, e.g. the PIN. The second type prompts the user to follow
instructions, e.g. to enter the PIN into the reader.
```

justin-stephenson replied:
Correct, but I am fine with prompt_init if it will be the initial prompt in each scenario.

ikerexxe replied:
I see your point, but I don't think that changes much so I'll keep it as it is

justin-stephenson commented on src/util/sss_pam_data.c:
```diff
@@ -237,3 +237,70 @@ pam_get_response_data(TALLOC_CTX *mem_ctx, struct pam_data *pd, int32_t type,
+    if (tmp_ctx == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_new failed.\n");
+        ret = ENOMEM;
+        goto done;
```
`goto done` tries to free NULL `tmp_ctx` here

ikerexxe replied:
That will have no impact, from the doc: Likewise, if "ptr" is NULL, then the function will make no modifications and returns -1.

In any case, I've change this part of the code to return `ENOMEM` directly in case of error.

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -186,9 +224,227 @@ obtain_prompts(struct confdb_ctx *cdb, TALLOC_CTX *mem_ctx,
         }
     }
 
+    if (sc_init_prompt == NULL) {
```
Do you need to check for an empty string here also? In `pc_list_add_smartcard()` you have 

```
    pc->data.smartcard.prompt_init = strdup(prompt_init != NULL ? prompt_init
                                            : "");
    if (pc->data.smartcard.prompt_init == NULL) {
        ret = ENOMEM;
        goto done;
    }
```

ikerexxe replied:
I think that this part of the code is more complex than necessary now that we only allow hardcoded strings. That's probably what is nailing you. Do you think it'd be wiser to assign the value with `talloc_strdup()` without checking whether the prompts are NULL? In all cases they are NULL so that wouldn't hurt.

justin-stephenson replied:
Okay this can be disregarded, I thought `sc_init_prompt` was coming from the value of ` pc->data.smartcard.prompt_init ` but I didn't check code outside of this PR in your passwordless_gdm branch.

ikerexxe replied:
Originally it came from another place, but then we changed our minds and we are hardcoding the value. I will change this to make the code clearer to understand

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -496,12 +496,16 @@ init_auth_data(TALLOC_CTX *mem_ctx, struct confdb_ctx *cdb,
 errno_t
-get_cert_names(TALLOC_CTX *mem_ctx, struct cert_auth_info *cert_list,
+get_cert_data(TALLOC_CTX *mem_ctx, struct cert_auth_info *cert_list,
                struct auth_data *_auth_data)
```
minor indentation offset

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -119,6 +120,79 @@ obtain_oauth2_data(TALLOC_CTX *mem_ctx, struct pam_data *pd,
+#ifdef BUILD_PASSKEY
+static errno_t
+obtain_passkey_data(TALLOC_CTX *mem_ctx, struct pam_data *pd,
+                   struct auth_data *_auth_data)
```
minor indentation issue

ikerexxe replied:
Fixed

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -30,6 +30,7 @@
 #include <string.h>
 
 #include "responder/pam/pamsrv.h"
+#include "responder/pam/pamsrv_passkey.h"
```
#ifdef BUILD_PASSKEY needed here?

ikerexxe replied:
Yes, most probably that's needed

justin-stephenson commented on src/tests/cmocka/test_pamsrv_json.c:
```diff
@@ -31,16 +31,16 @@
+#define SC_PIN_PROMPT       "Smartcard PIN"
+#define PASSKEY_INIT_PROMPT "Insert security key"
+#define PASSKEY_PIN_PROMPT  "Security key PIN"
+#define PASSKEY_TOUCH_PROMPT "Touch security key"
```
Is there a reason to differ from the default CLI prompts for passkey?

~~~
#define PASSKEY_DEFAULT_PIN_MSG     _("Enter PIN:")
~~~

ikerexxe replied:
I wasn't aware of this definition when I implemented the tests and nowadays I don't think it's worth changing anything when this is just a test and we want to make sure that the JSON message contains specific data.

justin-stephenson reviewed (changes requested):
I did a first round of review, some only minor comments are in-line. Overall it LGTM, I have 2 comments about the UX  - I only tested passkey, not smart card auth:

* Adding an IPA user with `ipa user-add pkuser --user-auth-type=passkey` then at the GDM login screen clicking the gear icon in the bottom right it shows `Password` and `Passkey` but I expect it would show `Passkey` only.

* Adding an IPA user with `ipa user-add pkuser --user-auth-type=passkey --user-auth-type=password`, at the GDM login screen clicking the gear icon in the bottom right it shows only `Password` under login methods, while the password prompt says "Insert your passkey device...". I would expect GDM login methods to show `Passkey` and `Password`.

justin-stephenson commented:
> 
>     * Adding an IPA user with `ipa user-add pkuser --user-auth-type=passkey --user-auth-type=password`, at the GDM login screen clicking the gear icon in the bottom right it shows only `Password` under login methods, while the password prompt says "Insert your passkey device...". I would expect GDM login methods to show `Passkey` and `Password`.

This was a config mistake on my part when `pam_json_services = gdm-switchable-auth` got removed from sssd.conf during my testing. Please disregard.

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -22,15 +22,32 @@
 
 #include "pamsrv_json.h"
 
+struct cert_auth_info {
```
I would suggest to move this into some header file because the same struct is currently used in src/sss_client/pam_sss.c and src/tests/cmocka/test_pamsrv_json.c. In src/responder/pam/pamsrv_p11.c a slightly different version is used, but maybe this can even be unified, but it does not have to be part of this PR.

ikerexxe replied:
Yeah, this is noted as an improvement but I'd like to change as few things as possible outside of the newly generated files.

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -186,9 +224,227 @@ obtain_prompts(struct confdb_ctx *cdb, TALLOC_CTX *mem_ctx,
     }
 
+    if (sc_init_prompt == NULL) {
+        sc_init_prompt = talloc_strdup(tmp_ctx, "Insert smartcard");
```
Here and below `DEFAULT_SMARTCARD_PROMPT_INIT` and `DEFAULT_SMARTCARD_PROMPT_PIN` can be used, respectively, if the defines are move into a header file.

ikerexxe replied:
I moved `DEFAULT_SMARTCARD_PROMPT_INIT` to the header file and left `DEFAULT_SMARTCARD_PROMPT_PIN` where it is. For the latter the text is not the same and it doesn't make sense to move it.

justin-stephenson reviewed (approved):
Changes LGTM, thank you (small note I primarily focused on the passkey part).

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -186,9 +224,227 @@ obtain_prompts(struct confdb_ctx *cdb, TALLOC_CTX *mem_ctx,
+        goto done;
+    }
+
+    //TODO: check also SSS_PAM_CERT_INFO_WITH_HINT?
```
The hint is only important for the "classical" GDM Smartcard behavior where GDM called PAM without a username and expected that the name is derived from the certificates on the Smartcard. In case the certificate(s) are mapped to multiple different users and hinting is enabled, there will be a second prompt besides the PIN prompt to enter a user name so that SSSD can use this name to decide how to move forward.

If I understand the new GDM workflow correctly a username will be always selected first and hence hinting is not needed.

ikerexxe replied:
I forgot about this `TODO`. Thank you for catching it. Removed.

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -301,6 +596,31 @@ json_format_priority(bool password_auth, bool oauth2_auth, json_t **_priority)
+                ret = ENOMEM;
+                goto done;
+            }
+            json_priority = json_string(key);
```
There is `json_sprintf()` which would make this shorter.

ikerexxe replied:
`json_sprintf()` relies on a hardcoded structure but we add authentication methods only if they are available for the user. In addition, this code changed a bit in a later commit and I don't think it makes much sense.

Feel free to reopen if you still think this applies.

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -119,6 +119,27 @@ obtain_oauth2_data(TALLOC_CTX *mem_ctx, struct pam_data *pd, char **_uri,
+        }
+        i++;
+    }
+
```
I guess something like `return ENOENT;` or similar is missing here.

ikerexxe replied:
I changed the code and this is no longer relevant

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -866,10 +887,33 @@ json_unpack_oauth2_code(TALLOC_CTX *mem_ctx, char *json_auth_msg,
 }
 
+errno_t
+json_unpack_smartcard(json_t *jroot, char **_pin)
```
Wouldn't it be more clear to use `const char **_pin` here (and for `json_unpack_passkey()` as well). This would better indicate that the caller does not has to free the string.

ikerexxe replied:
Done

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -944,6 +991,45 @@ json_unpack_auth_reply(struct pam_data *pd)
+                goto done;
+            }
+
+            cert_num = atoi(index);
```
`atoi()` does not do any proper error checking, better use one of our `strto*()` functions and check for errors becasuse `index` is coming from not validated input.

ikerexxe replied:
I removed this part of the code in a latter commit so no need to change it

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -944,6 +991,45 @@ json_unpack_auth_reply(struct pam_data *pd)
             goto done;
         }
+
+        if (strstr(key, "smartcard") != NULL) {
```
I would like to suggest to use `strncmp(key, "smartcard", 9` to compare only the beginning of `key` and not search the whole string. As an alternative `strstr(key, "smartcard") == key` might work as well, but I'm not sure.

ikerexxe replied:
Done

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -238,11 +343,38 @@ obtain_prompts(struct confdb_ctx *cdb, TALLOC_CTX *mem_ctx,
     }
 
+    if (passkey_init_prompt == NULL) {
+        passkey_init_prompt = talloc_strdup(tmp_ctx, "Insert security key");
```
Maybe here and in the following `PASSKEY_INIT_PROMPT`, `PASSKEY_PIN_PROMPT` and `PASSKEY_TOUCH_PROMPT` can be used if the defines are move into a header file?

ikerexxe replied:
Done

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -208,104 +208,17 @@ obtain_prompts(struct confdb_ctx *cdb, TALLOC_CTX *mem_ctx,
-                                CONFDB_PC_PASSWORD_PROMPT, "",
-                                &password_prompt);
-        if (ret != EOK) {
+        password_prompt = talloc_strdup(tmp_ctx, "Password");
```
In this version `password_prompt` will be always `NULL` since it cannot be set before, so the `NULL` check can be removed here and for the other prompts below as well.

It would be good to use a `#define` for "Password" (and for the other strings coming from the design). 

Will the decision "to leave them fixed and use the strings proposed by Allan in the mockups design." be a final one or will it be revisited at a later time to see how to make the prompts configurable? If this is final, why do the prompts have to be added to the JSON message at all? If there is a reason, e.g. other clients than GDM, they can be added as constants to the JSON strings in `json_format_mechanisms()` directly.

ikerexxe replied:
> In this version `password_prompt` will be always `NULL` since it cannot be set before, so the `NULL` check can be removed here and for the other prompts below as well.

Yes, I think I also discussed this topic with Justin in his review. Updated.

> It would be good to use a `#define` for "Password" (and for the other strings coming from the design).

I've moved the definition to `pamsrv_json.h`.

> Will the decision "to leave them fixed and use the strings proposed by Allan in the mockups design." be a final one or will it be revisited at a later time to see how to make the prompts configurable? If this is final, why do the prompts have to be added to the JSON message at all? If there is a reason, e.g. other clients than GDM, they can be added as constants to the JSON strings in `json_format_mechanisms()` directly.

It's unclear yet. We are still discussing this topic but as we want to deliver it as soon as possible I'm keeping the strings in the JSON message even though they can't be configured. Once it's clear what we'll do with this topic (this may take months) I'll revisit this topic and make the strings configurable or remove from the JSON message.

justin-stephenson commented:
Hi @ikerexxe Can you please re-upload SSSD build to your COPR https://copr.fedorainfracloud.org/coprs/ipedrosa/passwordles-gdm/ again? The rpms for latest build ` 2.11.1-5test ` are no longer available.

ikerexxe committed:
Responder: move DEFAULT_SMARTCARD_PROMPT_INIT to header

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: fix passkey auth when user-verification is off

When authenticating with a passkey, different PAM code paths within SSSD
can result in the `authtok` containing data even when the user did not
enter a PIN. Depending on the flow (e.g., triggered by `gdm` vs. `su`),
this data might be an empty string or non-printable characters like `^L`
(form feed).

The previous code had two issues:
1.  It only checked if the `authtok` was non-empty
    (`sss_authtok_get_type(...) != SSS_AUTHTOK_TYPE_EMPTY`). If user
    verification was disabled, this check would incorrectly pass for
    these 'junk' `authtok` values. This caused SSSD to prepare and send
    an erroneous PIN to the passkey helper.

2.  In the case where the `authtok` *was* correctly empty, the check
    would fail, `write_buf_len` would remain 0, and the `if
    (write_buf_len != 0)` block containing the `write_pipe_send` call
    would be skipped. This stalled the authentication flow, as the
    callback to continue the process was never set.

This patch fixes both issues:
1.  The `user_verification` setting is now stored in the state struct.
    The logic is updated to only prepare the PIN buffer if the `authtok`
    is non-empty *and* user verification is required
    (`state->user_verification != PAM_PASSKEY_VERIFICATION_OFF`).

2.  The `write_pipe_send` call is moved outside the conditional block so
    it always runs. This ensures that the asynchronous child
    communication (via `passkey_child_write_done`) is always triggered,
    even if the write buffer is empty (0-length).

This resolves both failure modes: junk PINs are no longer sent when
verification is off, and the auth flow no longer stalls when no PIN is
present.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe commented:
Let me summarize the latest changes I pushed:
* Address Sumit's comments.
* Remove a commit containing debug messages.
* Add a commit to move DEFAULT_SMARTCARD_PROMPT_INIT to a header file (penultimate commit).
* Fix passkey auth when user-verification is off (last commit).

justin-stephenson commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -1128,8 +1130,9 @@ static errno_t passkey_child_exec(struct tevent_req *req)
-        if (sss_authtok_get_type(state->pd->authtok) != SSS_AUTHTOK_TYPE_EMPTY) {
+        /* PIN is needed only when user verification is required */
+        if (sss_authtok_get_type(state->pd->authtok) != SSS_AUTHTOK_TYPE_EMPTY
+            && state->user_verification == PAM_PASSKEY_VERIFICATION_ON) {
```
this will break passkey local auth when passkey verification is 'unset' here as the passkey-child will then enforce user-verification and hence require a PIN.

```
(2025-10-29 10:02:24): [passkey_child[963205]] [get_device_options] (0x0100): Policy didn't indicate any preference for user-verification but the key settings are enforcing it. Thus, enforcing the user-verification.
(2025-10-29 10:02:24): [passkey_child[963205]] [authenticate] (0x0400): Resetting assert options.
(2025-10-29 10:02:24): [passkey_child[963205]] [authenticate] (0x0400): Resetting assert client data.
(2025-10-29 10:02:24): [passkey_child[963205]] [authenticate] (0x0400): Decoding public key.
(2025-10-29 10:02:24): [passkey_child[963205]] [authenticate] (0x0400): Getting assert.
(2025-10-29 10:02:24): [passkey_child[963205]] [passkey_recv_pin] (0x0020): Missing PIN.
```


Can you change to:

       if (sss_authtok_get_type(state->pd->authtok) != SSS_AUTHTOK_TYPE_EMPTY
            && (state->user_verification == PAM_PASSKEY_VERIFICATION_ON || 
            state->user_verification == PAM_PASSKEY_VERIFICATION_OMIT)) {

or `user_verification != PAM_PASSKEY_VERIFICATION_OFF` if you prefer.

ikerexxe replied:
Is that the reason why automatic tests are failing? I was unable to reproduce them with my local environment...

ikerexxe replied:
This is fixed now

ikerexxe commented:
CI failures seem to be unrelated so you can proceed with a new round of review.

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -589,6 +1214,55 @@ json_unpack_auth_reply(struct pam_data *pd)
+                    goto done;
+                }
+            } else {
+                ret = sss_authtok_set_local_passkey_pin(pd->authtok, pin);
```
I think `sss_authtok_set_passkey_krb()` and `sss_authtok_set_local_passkey_pin()` should be consistent in either requiring a `user_verification` argument or not.

ikerexxe replied:
The underlying APIs, `sss_auth_pack_passkey_blob()` and `sss_authtok_set_string()` aren't consistent in that behaviour. I could require, but then we'd be adding additional arguments for a function that doesn't need them

justin-stephenson replied:
okay I don't insist.

justin-stephenson commented on src/responder/pam/pamsrv_json.h:
```diff
@@ -30,64 +30,126 @@
+struct passkey_data {
+    bool enabled;
+    char *init_prompt;
+    bool key_connected;
```
How is key_connected used?

ikerexxe replied:
Currently it isn't. I added it to make the integration with https://github.com/SSSD/sssd/pull/7983 easier

justin-stephenson replied:
I don't remember using this in that PR but i'll defer to your judgement.

ikerexxe replied:
`key_connected` is `obtained` from `pam_preflight_data`, but with a different name

justin-stephenson replied:
This can be resolved, thanks.

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -92,103 +112,522 @@ obtain_oauth2_data(TALLOC_CTX *mem_ctx, struct pam_data *pd, char **_uri,
+    return ret;
+}
+
+#ifdef BUILD_PASSKEY
```
Just to be sure have you tested building your code without `--with-passkey` ?

ikerexxe replied:
I tested that it builds, I didn't do any functional test. I'm expecting that both features are enabled together

justin-stephenson replied:
Okay thanks.

justin-stephenson commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -92,103 +112,522 @@ obtain_oauth2_data(TALLOC_CTX *mem_ctx, struct pam_data *pd, char **_uri,
+
+#ifdef BUILD_PASSKEY
+static errno_t
+obtain_passkey_data(TALLOC_CTX *mem_ctx, struct pam_data *pd,
```
Certainly not necessary as part of this PR, but it would be nice if existing function `pam_eval_passkey_response()` could be made to use `pam_get_response_data()` how you do here, or maybe even `obtain_passkey_data` and `pam_eval_passkey_response` could be converged into a single function as they both are quite similar.

ikerexxe replied:
yeah, they aren't exactly the same but very similar so we could merge them, or at least make them depend on the same functions.

I would prefer to address that outside of this PR, as I would like to keep conflicting and risky code to a minimum. We can make this improvement later

justin-stephenson commented:
In my latest tests with  `sssd-client-2.11.1-11test.fc42.x86_64`:


- GDM login using IPA with PIN set: `Success`
- GDM login using IPA with no PIN set: `Success`
- GDM login using LDAP (local auth) with PIN set: `Success` Note: Fails with SELinux enabled (we already have a ticket open for this)
- GDM login using LDAP (local auth) with no PIN set: `Fails`, but fails on the command-line also. I plan to fix this in a separate PR.

ikerexxe committed:
Responder: add `gdm-switchable-auth` to `pam_p11_allowed_services`
defaults

The `pam_p11_allowed_services` option now includes `gdm-switchable-auth`
as one of the default allowed PAM services for smartcard authentication.
The service was added alongside the other GDM-related services
(gdm-smartcard and gdm-password) for logical grouping.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

sumit-bose commented on src/providers/krb5/krb5_child.c:
```diff
@@ -1195,49 +1183,168 @@ static krb5_error_code answer_password(krb5_context kctx,
+        if (strcmp(question_list[c], SSSD_IDP_OAUTH2_QUESTION) == 0) {
+            kerr = request_idp_oauth2(ctx, kr, rctx, _oath2_data);
+        } else if (strcmp(question_list[c], SSSD_PASSKEY_QUESTION) == 0) {
+            kerr = request_passkey(ctx, kr, rctx);
```
Hi,

this is currently incomplete and would break 2FA and Smartcard authentication. All authentication types must be handled here and the corresponding PAM response data must be set. To make Smartcard authentication work I added
```
diff --git a/src/providers/krb5/krb5_child.c b/src/providers/krb5/krb5_child.c
index c49265b6f..f199b2c3c 100644
--- a/src/providers/krb5/krb5_child.c
+++ b/src/providers/krb5/krb5_child.c
@@ -1247,6 +1247,13 @@ static krb5_error_code sss_krb5_auth_methods_request(krb5_context ctx,
             kerr = request_idp_oauth2(ctx, kr, rctx, _oath2_data);
         } else if (strcmp(question_list[c], SSSD_PASSKEY_QUESTION) == 0) {
             kerr = request_passkey(ctx, kr, rctx);
+        } else if (strcmp(question_list[c], KRB5_RESPONDER_QUESTION_PKINIT) == 0) {
+            DEBUG(SSSDBG_TRACE_ALL, "Setting pkinit_prompting.\n");
+            kerr = pam_add_response(kr->pd, SSS_CERT_AUTH_PROMPTING, 0,
+                                   NULL);
+            if (kerr != EOK) {
+                DEBUG(SSSDBG_CRIT_FAILURE, "Failed to added pkinit prompting data.\n");
+            }
         }
 
         if (kerr == EOK) {
```
but of course it would be better to have a dedicated `request_*` function which basically contains what was run in the corresponding original `answer_*` during `SSS_PAM_PREAUTH`. Additionally `pam_add_prompting()` should be  called in `k5c_send_and_recv()` if the prompting flags are used or the new `request_*` functions can call `pam_add_response()` directly.

Jfyi, to test Smartcard authentication you can use Vema's https://github.com/SSSD/sssd/pull/8159. It is not merged yet but applies cleanly on top of your patches.

bye,
Sumit

ikerexxe replied:
I tackled this in the last commit of the series. Please take a look at it. For additional information please check https://github.com/ikerexxe/sssd/pull/12#issuecomment-3510938611

sumit-bose replied:
Hi,

thanks, this looks good now.

bye,
Sumit

ikerexxe commented:
Major changes in the latest version:
* Addition of `gdm-switchable-auth` to `pam_p11_allowed_services` defaults (Third-to-last commit)
* Setting `gdm-switchable-auth` as default for `pam_json_services` (Penultimate commit)
* Setting PAM response data in krb5_child.c (Last commit, and once approved I'll squash it)

Apart from that I also run the test at https://github.com/SSSD/sssd/pull/8159 locally to make sure that smartcard auth is working. Finally, build is available in COPR repo.

ikerexxe committed:
sss_client: prevent JSON auth during password change preauth

When a user's password expires after successful JSON authentication, the
fallback to traditional password change fails. Add
PAM_CLI_FLAGS_CHAUTHTOK_PREAUTH flag to distinguish password change
preauth from normal authentication preauth. When this flag is set, the
PAM responder skips JSON message generation and returns traditional
preauth data instead.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -92,103 +112,522 @@ obtain_oauth2_data(TALLOC_CTX *mem_ctx, struct pam_data *pd, char **_uri,
+        goto done;
+    }
+
+    ret = get_cert_list(mem_ctx, pd, &cert_list);
```
Hi,

if the user has inserted a Smartcard or a Yubikey with certificates on it there will always be certificates returned here even if the server-side does not support Smartcard authentication. You have to call `pam_get_auth_types()` or pass `struct pam_resp_auth_type` down here and check if `cert_auth` is set.

Similar with `( *_auth_data)->pswd->enabled = true;` because currently password authentication is enabled unconditionally even if the server-side does not support it.

bye,
Sumit

sumit-bose replied:
Btw, I've written a small extension to `sssctl user-checks` to help me test the patches at https://github.com/sumit-bose/sssd/commit/1635b33d94b87f4fde765c0c26593ce74977e53b .

If you call `sssctl user-checks testuser -s gdm-switchable-auth -a auth` it will print the JSON data GDM will receive.

HTH

bye,
Sumit

ikerexxe replied:
I knew a function like that must exist but I was unable to find it when I was writing the implementation. I've update the code to use `pam_get_auth_types()` instead.

ikerexxe committed:
Responder: change authentication mechanism detection

Use `pam_get_auth_types()` to detect the available mechanisms for a
user.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe commented:
Major changes in the latest version:
* Removed setting `gdm-switchable-auth` as default for `pam_json_services`. In a conversation with Sumit he mentioned that we may break 2FA by enabling this by default, so I decided to remove it.
* Responder: change authentication mechanism detection to `pam_get_auth_types()`

This is ready for review again ;)

sumit-bose commented:
Hi,

thank you for the last update, controlling the available authentication types work much better now.

I would suggest to mention the limitation of the JSON protocol with respect to two-factor authentication and the man page entry of `pam_json_services`. Additionally, the related man page section is not build at all because `JSON_PAM_CONDS` is missing in the `CONDS = ...` line in `src/man/Makefile.am` already in the older patch set.

bye,
Sumit

ikerexxe committed:
man: clarify and fix `pam_json_services` compilation

Add a note to clarify that 2FA isn't supported in JSON protocol and fix
man page compilation for `pam_json_services` option.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe commented:
> I would suggest to mention the limitation of the JSON protocol with respect to two-factor authentication and the man page entry of `pam_json_services`. Additionally, the related man page section is not build at all because `JSON_PAM_CONDS` is missing in the `CONDS = ...` line in `src/man/Makefile.am` already in the older patch set.

Updated in a new commit (last one).

You can disregard CI failures, Fedora server doesn't seem to respond to the `get_matrix.py` script. I guess this is cloudflare related and will rerun CI tomorrow

sumit-bose commented:
Hi,

thanks for the updates, (so far) I have no further comments.

bye,
Sumit

ikerexxe commented:
CI failures seem to be unrelated so I'm merging.
